### PR TITLE
Add driver for RC2014 Mini

### PR DESCRIPTION
**Pull request by cpswan**
Driver for [RC2014 Mini](https://rc2014.co.uk/full-kits/rc2014-mini/) based on [gsz80.cpp](https://github.com/mamedev/mame/blob/master/src/mame/drivers/gsz80.cpp) from #6643

### CODE REVIEW

**Comment by Robbbert** [comment]
mame.lst - carriage return at end of file is missing.

mess.flt, mess.lua changes haven't been included in the PR

**Comment by startaq** [comment]
If this is just a copy of gsz80.cpp then you can just add a new rom section to  it.

**Comment by cpswan** [comment]
@Robbbert [b1c9aae](https://github.com/mamedev/mame/pull/9331/commits/b1c9aae015d52c6047496168e569a2156371165e) should fix the points that you raise.

@startaq the ROM for RC2014 56k BASIC is the same one as used by gsz80.cpp, as the RC2014 modular design is just Grant Searle's system split into modules; then the Mini brings together the core modules onto a single board.

**Comment by cpswan** [comment]
@startaq I've switched the ROM to the RC2014 specific Small Computer Monitor (SCM), and in doing that realised that I shouldn't have been using 56k BASIC at all, as the RC2014 Mini has 32K RAM, so I've fixed that in the memory map.

I can add the 32K BASIC ROM if you can point me at a good simple example of where multiple ROMs are defined for a driver?

**Comment by startaq** [comment]
Well my point was that the drivers seem largely identical, so it might make sense to put both into the same file.

**Comment by cpswan** [comment]
@startaq it's all in gsz80.cpp now.

**Review by cuavas — CHANGES_REQUESTED** [review]

File: src/mame/drivers/gsz80.cpp [hunk 1]
```diff
@@ -38,20 +41,55 @@ class gsz80_state : public driver_device
...
 	map.unmap_value_high(); // unmapped addresses return 0xff
 	map(0x80, 0xbf).rw("acia", FUNC(acia6850_device::read), FUNC(acia6850_device::write));
 }
 
+void rc2014mini_state::rc2014mini_io(address_map &map)
+{
+	map.global_mask(0xff);  // use 8-bit ports
+	map.unmap_value_high(); // unmapped addresses return 0xff
+	map(0x80, 0xbf).rw("acia", FUNC(acia6850_device::read), FUNC(acia6850_device::write));
+}
```
[the remark this answers is not in the document]
Comment (cpswan) [reply src/mame/drivers/gsz80.cpp]: removed

File: src/mame/drivers/gsz80.cpp [hunk 2]
```diff
@@ -38,20 +41,55 @@ class gsz80_state : public driver_device
 	required_device<acia6850_device> m_acia;
 };
 
+class rc2014mini_state : public gsz80_state
+{
+public:
+	rc2014mini_state(const machine_config &mconfig, device_type type, const char *tag)
+		: gsz80_state(mconfig, type, tag)
+		, m_maincpu(*this, "maincpu")   // Tag name for Z80 is "maincpu"
+		, m_acia(*this, "acia")         // Tag name for UART is "acia"
```
[the remark this answers is not in the document]
Comment (cpswan) [reply src/mame/drivers/gsz80.cpp]: done, and class simplified

File: src/mame/drivers/gsz80.cpp [hunk 3]
```diff
@@ -89,12 +127,50 @@ void gsz80_state::gsz80(machine_config &config)
...
+	acia_clock.signal_handler().set("acia", FUNC(acia6850_device::write_txc));
+	acia_clock.signal_handler().append("acia", FUNC(acia6850_device::write_rxc));
+
+	// Configure a "default terminal" to connect to the 6850, so we have a console
+	rs232_port_device &rs232(RS232_PORT(config, "rs232", default_rs232_devices, "terminal"));
+	rs232.rxd_handler().set(m_acia, FUNC(acia6850_device::write_rxd));
+	rs232.dcd_handler().set(m_acia, FUNC(acia6850_device::write_dcd));
+	rs232.cts_handler().set(m_acia, FUNC(acia6850_device::write_cts));
+	rs232.set_option_device_input_defaults("terminal", DEVICE_INPUT_DEFAULTS_NAME(terminal)); // must be below the DEVICE_INPUT_DEFAULTS_START block
+}
```
[the remark this answers is not in the document]
Comment (cpswan) [reply src/mame/drivers/gsz80.cpp]: done

**Comment by happppp** [comment]
The hardware specs you're emulating match this: http://searle.x10host.com/z80/SimpleZ80_32K.html
"rc2014mini_32k.bin" is the same as ROM_32K.HEX from there.

RF2014 Mini has a 512x8 ROM instead, with 3 jumpers for bankswitching, see schematics: https://rc2014.co.uk/full-kits/rc2014-mini/

**Comment by cpswan** [comment]
@happppp I'm not surprised that @RFC2795 got the 32K BASIC 'R' ROM from Grant Searle, given that the 56K BASIC 'K' ROM is from Grant. As I said in an earlier comment:
> the RC2014 modular design is just Grant Searle's system split into modules; then the Mini brings together the core modules onto a single board

As the jumpers on the RC2014 Mini are for selecting 8K pages from a 64K ROM it seemed like the more MAMEish way to do things would be to present those 8K pages as selectable BIOSes (in this case selected in MAME rather than by setting jumpers). Though I can see a counter argument that jumpers would be more authentic to the original hardware.

6 of the 8 pages in the factory ROM (labelled R0000009) are empty, which offers the possibility of using the space for different firmware, but there's nothing interesting there out of the box for the various intermediate jumper settings.

**Comment by cpswan** [comment]
@happppp thinking about it some more, I guess the MAME way is to be as authentic as possible to the original hardware (and the way that ROMs are presented). Can you please point me at a driver or two that uses jumpers for bank selection from a ROM, and I'll have a go at implementing the A13/14/15 jumpers on the Mini?

**Comment by happppp** [comment]
Off the top of my head, no I don't know a driver that bankswitches the program rom with jumpers.
You can do it with these: PORT_CONFNAME/PORT_CONFSETTING, INPUT_CHANGED_MEMBER, required_memory_bank

**Comment by cpswan** [comment]
@happppp Latest push uses jumpers to select between BASIC and SCM, and is making use of the full r0000009.bin ROM.

**Comment by cpswan** [comment]
@happppp the last few commits should take care of your earlier comments. Thanks for being so patient as I blunder around with this. "We do these things not because they are easy, but because we thought they would be easy...". I've learned a lot more than I expected to on this journey so far, but I'm still not a minimally competent C++ programmer, and I'm still borrowing too heavily from old drivers (and snippets from forks people have pointed me at) that aren't up to scratch with current conventions.

File: src/mame/drivers/gsz80.cpp [hunk 4]
```diff
@@ -89,12 +161,32 @@ void gsz80_state::gsz80(machine_config &config)
...
+{
+	gsz80(config);
+	m_maincpu->set_addrmap(AS_PROGRAM, &rc2014mini_state::rc2014mini_mem);
+}
+
+static INPUT_PORTS_START( rc2014mini )
+	PORT_START("A13-15")   /* jumpers to select ROM region */
+	PORT_DIPNAME( 0x7, 0, "ROM" )
+	PORT_DIPSETTING( 0, "BASIC" )
+	PORT_DIPSETTING( 0x7, "SCM" )
```
[the remark this answers is not in the document]
Comment (cpswan) [reply src/mame/drivers/gsz80.cpp]: done in [aa0a830](https://github.com/mamedev/mame/pull/9331/commits/aa0a830d771db9fe890b330cc44d664de096ab00)

**Comment by happppp** [comment]
optional_ioport can be changed to required
m_region_maincpu can be removed as it's unused.
void update_banks(); as well

Is "ROM" a good name for that CONFNAME? If you want it to be brief, "ROM Bank" describes it better.
Is that github link to that rom folder ok to put in our source? I thought these files were free to use but now I see Microsoft Basic in there and I'm not sure anymore.

**Comment by cpswan** [comment]
@happppp most recent comments dealt with

File: src/mame/drivers/gsz80.cpp [hunk 5]
```diff
@@ -38,13 +41,54 @@ class gsz80_state : public driver_device
...
+// Set ROM bank from machine CONF at Reset
+void rc2014mini_state::machine_reset()
+{
+	m_rombank->set_entry(m_jump_rom->read() & 7);
+}
+
+// RC2014 Mini only has 32K RAM
+void rc2014mini_state::rc2014mini_mem(address_map &map)
+{
+	map(0x0000, 0x1fff).rom().bankr("bank");
```
Comment (cuavas) [inline src/mame/drivers/gsz80.cpp:88]: You shouldn’t have the `.rom()` in there – the `.bankr("bank")` is overriding it.  It’s confusing like this.  It should just have the `.bankr("bank")`

**Review by cuavas — CHANGES_REQUESTED** [review]
I think that’s the last issue if @happppp doesn’t notice anything else.

**Pull request merged**